qualtran/serialization/data_types.py

```diff
@@ -13,40 +13,56 @@
 #  limitations under the License.
 from typing import Union
 
+import sympy
+
 from qualtran import BoundedQUInt, QAny, QBit, QDType, QFxp, QInt, QIntOnesComp, QUInt
 from qualtran.protos import data_types_pb2
 from qualtran.serialization.args import int_or_sympy_from_proto, int_or_sympy_to_proto
 
-SerializedDataTypes = Union[
-    data_types_pb2.QBit,
-    data_types_pb2.QAny,
-    data_types_pb2.QInt,
-    data_types_pb2.QIntOnesComp,
-    data_types_pb2.QUInt,
-    data_types_pb2.BoundedQUInt,
-    data_types_pb2.QFxp,
-]
+
+def int_or_sympy_from_qdt_proto(val: data_types_pb2.QDataType) -> Union[int, sympy.Expr]:
+    if val.HasField('qbit'):
+        raise ValueError('QBit case should have been handled already.')
+    if val.HasField('qany'):
+        return int_or_sympy_from_proto(val.qany.bitsize)
+    if val.HasField('qint'):
+        return int_or_sympy_from_proto(val.qint.bitsize)
+    if val.HasField('qint_ones_comp'):
+        return int_or_sympy_from_proto(val.qint_ones_comp.bitsize)
+    if val.HasField('quint'):
+        return int_or_sympy_from_proto(val.quint.bitsize)
+    if val.HasField('bounded_quint'):
+        return int_or_sympy_from_proto(val.bounded_quint.bitsize)
+    if val.HasField('qfxp'):
+        return int_or_sympy_from_proto(val.qfxp.bitsize)
+    raise ValueError(f"Cannot deserialize {val=}")
```

Comment:
I think this helper is moot and should be removed, the name and interface is confusing and it's not saving us much since we are doing all the same isinstance checks in the `data_type_to_proto` function below.

Reply:
Sorry I was a bit too eager to hit merge: https://github.com/quantumlib/Qualtran/pull/698